cdap-master/src/main/java/co/cask/cdap/data/runtime/main/MasterServiceMain.java

```diff
@@ -606,6 +609,14 @@ private TwillController startTwillApplication() {
           preparer.addSecureStore(secureStoreUpdater.update(null, null));
         }
 
+        // add hadoop classpath to application classpath and exclude hadoop classes from bundle jar.
+        String yarnAppClassPath = hConf.get(YarnConfiguration.YARN_APPLICATION_CLASSPATH,
+                                            Joiner.on(",").join(YarnConfiguration.DEFAULT_YARN_APPLICATION_CLASSPATH));
+
+        List<String> applicationClassPaths = Lists.newArrayList();
+        Iterables.addAll(applicationClassPaths, Splitter.on(",").trimResults().split(yarnAppClassPath));
+        preparer.withApplicationClassPaths(applicationClassPaths).withBundlerClassAcceptor(new HadoopClassExcluder());
```

Comment:
withApplicationClassPaths takes `Iterables`, hence you don't need to copy to a `List`. Just do `withApplicationClassPaths(Splitter.on(",")...split(...))` should be fine.